android-sdk/src/main/java/co/elastic/apm/android/sdk/internal/otel/ElasticClock.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *	http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.android.sdk.internal.otel;
+
+import androidx.annotation.VisibleForTesting;
+
+import java.util.concurrent.TimeUnit;
+
+import co.elastic.apm.android.common.internal.logging.Elog;
+import co.elastic.apm.android.sdk.internal.time.SystemTimeProvider;
+import co.elastic.apm.android.sdk.internal.time.ntp.TrueTimeWrapper;
+import io.opentelemetry.sdk.common.Clock;
+
+public class ElasticClock implements Clock {
+    private final TrueTimeWrapper trueTimeWrapper;
+    private final SystemTimeProvider systemTimeProvider;
+
+    @VisibleForTesting
+    public ElasticClock(TrueTimeWrapper trueTimeWrapper, SystemTimeProvider systemTimeProvider) {
+        this.trueTimeWrapper = trueTimeWrapper;
+        this.systemTimeProvider = systemTimeProvider;
+    }
+
+    public ElasticClock(TrueTimeWrapper trueTimeWrapper) {
+        this(trueTimeWrapper, SystemTimeProvider.get());
+    }
+
+    @Override
+    public long now() {
+        if (trueTimeWrapper.isInitialized()) {
```

Comment:
can we end up with negative durations if this condition changes in the middle of a span? Or is the nanoTime() function always used for the span duration? Just wondering.

Reply:
That's my understanding, too.

Reply:
I've confirmed that they use nanoTime to calculate the span duration when the [span is ended](https://github.com/open-telemetry/opentelemetry-java/blob/84170709c4d9ee0bb4c923e2fda41fd107645c4a/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/SdkSpan.java#L425) (they call [AnchoredClock.now()](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/AnchoredClock.java#L50) which in turn calls Clock.nanoTime). I've also covered that case with [this test](https://github.com/elastic/apm-agent-android/pull/67/files#diff-d67c60a8ebb8521a0586bae0e5f653837491ba7ff209fc99803a938ac77286c9R49) just to make sure it stays the same in further OTel releases.